File: 4-Infrastructure/infra/search_adapter.py
#!/usr/bin/env python3
"""
search_adapter.py — Search implementation for Research Stack

Integrates multiple search providers (Google Surf, Brave Search) 
into a unified interface.
"""

import os
import json
import subprocess
from typing import List, Dict, Any, Optional

class SearchResult:
    def __init__(self, title: str, url: str, snippet: str):
        self.title = title
        self.url = url
        self.snippet = snippet

    def to_dict(self):
        return {"title": self.title, "url": self.url, "snippet": self.snippet}

class SearchProvider:
    def search(self, query: str, limit: int = 5) -> List[SearchResult]:
        raise NotImplementedError

class GoogleSurfProvider(SearchProvider):
    def search(self, query: str, limit: int = 5) -> List[SearchResult]:
        # Using npx to call the google-surf-mcp tool if possible
        # Actually, since it's an MCP server, we might prefer a direct search if it has a CLI
        # But for now, we'll simulate or use a fallback if not configured.
        print(f"Searching Google Surf for: {query}")
        # Placeholder for real integration
        return []

class BraveSearchProvider(SearchProvider):
    def __init__(self, api_key: Optional[str] = None):
        self.api_key = api_key or os.getenv("BRAVE_API_KEY")

    def search(self, query: str, limit: int = 5) -> List[SearchResult]:
        if not self.api_key or "YOUR_BRAVE_API_KEY" in self.api_key:
            return []
            
        import requests
        url = "https://api.search.brave.com/res/v1/web/search"
        headers = {
            "Accept": "application/json",
            "Accept-Encoding": "gzip",
            "X-Subscription-Token": self.api_key
        }
        params = {"q": query, "count": limit}
        
        try:
            resp = requests.get(url, headers=headers, params=params, timeout=10)
            resp.raise_for_status()
            data = resp.json()
            results = []
            for item in data.get("web", {}).get("results", []):
                results.append(SearchResult(item["title"], item["url"], item["description"]))
            return results
        except Exception as e:
            print(f"Brave Search Error: {e}")
            return []

class UnifiedSearcher:
    def __init__(self):
        self.providers = [
            BraveSearchProvider(),
            GoogleSurfProvider()
        ]

    def search(self, query: str, limit: int = 5) -> List[SearchResult]:
        for provider in self.providers:
            results = provider.search(query, limit)
            if results:
                return results
        return []

File: 4-Infrastructure/infra/ene-session-sync/src/search_adapter.rs
#![allow(dead_code)]
//! search_adapter.rs — Web search provider adapters.
//!
//! Port of search_adapter.py (77 lines).
//! Two providers:
//!   - `GoogleSurfProvider` — placeholder (real dispatch via MCP server)
//!   - `BraveSearchProvider` — live search via Brave Search API

use serde_json::{json, Value};

// ─────────────────────────────────────────────────────────────────────────────
// §1  SearchResult
// ─────────────────────────────────────────────────────────────────────────────

/// A single web search result.
#[derive(Debug, Clone)]
pub struct SearchResult {
    pub title: String,
    pub url: String,
    pub snippet: String,
}

impl SearchResult {
    pub fn new(title: impl Into<String>, url: impl Into<String>, snippet: impl Into<String>) -> Self {
        Self {
            title: title.into(),
            url: url.into(),
            snippet: snippet.into(),
        }
    }

    /// Serialize to a JSON object with `title`, `url`, and `snippet` keys.
    pub fn to_value(&self) -> Value {
        json!({
            "title":   self.title,
            "url":     self.url,
            "snippet": self.snippet,
        })
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// §2  GoogleSurfProvider
// ─────────────────────────────────────────────────────────────────────────────

/// Placeholder Google search provider.
///
/// Real integration is handled via an MCP server external to this binary.
/// Returns an empty result set so call-sites compile and type-check cleanly.
pub struct GoogleSurfProvider;

impl GoogleSurfProvider {
    pub fn new() -> Self {
        Self
    }

    /// Submit a search query.  Always returns an empty vec — see module doc.
    pub async fn search(&self, _query: &str, _limit: usize) -> Vec<SearchResult> {
        // Dispatch is handled by the MCP server layer.  This shim exists so
        // that code depending on `GoogleSurfProvider` compiles and can be
        // wired up later without API changes.
        Vec::new()
    }
}

impl Default for GoogleSurfProvider {
    fn default() -> Self {
        Self::new()
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// §3  BraveSearchProvider
// ─────────────────────────────────────────────────────────────────────────────

/// Live web search via the Brave Search REST API.
///
/// Requires a `BRAVE_API_KEY` subscription token.  Returns an empty result set
/// when no key is configured or when the API call fails.
pub struct BraveSearchProvider {
    api_key: Option<String>,
    http: reqwest::Client,
}

impl BraveSearchProvider {
    pub fn new(api_key: Option<String>) -> Self {
        Self {
            api_key,
            http: reqwest::Client::new(),
        }
    }

    /// Build from the `BRAVE_API_KEY` environment variable.
    pub fn from_env() -> Self {
        Self::new(std::env::var("BRAVE_API_KEY").ok())
    }

    /// Execute a web search query and return up to `limit` results.
    ///
    /// Hits `https://api.search.brave.com/res/v1/web/search`.
    /// Returns an empty vec when no API key is present or on any error.
    pub async fn search(&self, query: &str, limit: usize) -> Vec<SearchResult> {
        let key = match &self.api_key {
            Some(k) => k.clone(),
            None => return Vec::new(),
        };

        let url = "https://api.search.brave.com/res/v1/web/search";
        let resp = match self
            .http
            .get(url)
            .header("Accept", "application/json")
            .header("X-Subscription-Token", &key)
            .query(&[("q", query), ("count", &limit.to_string())])
            .send()
            .await
        {
            Ok(r) => r,
            Err(e) => {
                tracing::warn!("brave search request failed: {}", e);
                return Vec::new();
            }
        };

        if !resp.status().is_success() {
            tracing::warn!("brave search returned {}", resp.status());
            return Vec::new();
        }

        let body: Value = match resp.json().await {
            Ok(v) => v,
            Err(e) => {
                tracing::warn!("brave search response parse error: {}", e);
                return Vec::new();
            }
        };

        // Response schema: { "web": { "results": [ { "title", "url", "description" } ] } }
        let raw_results = body
            .get("web")
            .and_then(|w| w.get("results"))
            .and_then(|r| r.as_array())
            .cloned()
            .unwrap_or_default();

        raw_results
            .into_iter()
            .map(|item| {
                let title = item
                    .get("title")
                    .and_then(|t| t.as_str())
                    .unwrap_or("")
                    .to_string();
                let url = item
                    .get("url")
                    .and_then(|u| u.as_str())
                    .unwrap_or("")
                    .to_string();
                let snippet = item
                    .get("description")
                    .and_then(|d| d.as_str())
                    .unwrap_or("")
                    .to_string();
                SearchResult { title, url, snippet }
            })
            .collect()
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// §4  Tests
// ─────────────────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn search_result_to_value() {
        let r = SearchResult::new("My Title", "https://example.com", "A snippet.");
        let v = r.to_value();
        assert_eq!(v["title"], "My Title");
        assert_eq!(v["url"], "https://example.com");
        assert_eq!(v["snippet"], "A snippet.");
    }

    #[test]
    fn brave_no_key_returns_empty() {
        // Without an API key the provider is constructed correctly but will
        // return empty — we verify the struct, not the async path.
        let provider = BraveSearchProvider::new(None);
        assert!(provider.api_key.is_none());
    }

    #[test]
    fn google_surf_provider_default() {
        let _ = GoogleSurfProvider::default();
    }

    #[tokio::test]
    async fn google_surf_returns_empty() {
        let p = GoogleSurfProvider::new();
        let results = p.search("test query", 10).await;
        assert!(results.is_empty());
    }
}
